pkg/cluster/internal/create/actions/kubeadminit/init.go

```diff
@@ -106,14 +107,31 @@ func (a *action) Execute(ctx *actions.ActionContext) error {
 		}
 	}
 
-	// if we are only provisioning one node, remove the master taint
+	// if we are only provisioning one node, remove the control plane taint
 	// https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/#master-isolation
 	if len(allNodes) == 1 {
+		// TODO: Once kubeadm 1.23 is no longer supported remove the <1.24 handling.
+		// TODO: Remove only the "control-plane" taint for kubeadm >= 1.25.
+		// https://github.com/kubernetes-sigs/kind/issues/1699
+		rawVersion, err := nodeutils.KubeVersion(node)
+		if err != nil {
+			return errors.Wrap(err, "failed to get Kubernetes version from node")
+		}
+		kubeVersion, err := version.ParseSemantic(rawVersion)
+		if err != nil {
+			return errors.Wrap(err, "could not parse Kubernetes version")
+		}
+		taints := []string{"node-role.kubernetes.io/control-plane-", "node-role.kubernetes.io/master-"}
+		if kubeVersion.LessThan(version.MustParseSemantic("v1.24.0")) {
+			taints = []string{"node-role.kubernetes.io/master-"}
+		}
```

Comment:
here is my confusion

if kubernetes > 1.24
it removes `"node-role.kubernetes.io/control-plane-", "node-role.kubernetes.io/master-` , should not remove `node-role.kubernetes.io/control-plane-` only?

the way I understand it is that we should have 3 cases,
1. > 1.24 remove `node-role.kubernetes.io/control-plane-` 
2. 1.20-1.24 remove `"node-role.kubernetes.io/control-plane-", "node-role.kubernetes.io/master-"`
3. < 1.20 remove `"node-role.kubernetes.io/master-"`

Reply:
note that taints and labels are staged differently.
the description of https://github.com/kubernetes/kubernetes/pull/107533 has the details.

my response here is about labels:
https://github.com/kubernetes-sigs/kind/pull/2595#issuecomment-1017451427

for taints and single CP nodes:
- in < 1.24 we want to remove only the `master` taint, because old versions of kubeadm only add the `master` taint.
- in >= 1.24 we want to remove the `master` and `control-plane` taints because kubeadm 1.24+ will start adding both taints.
- in 1.25 we have to touch kind again to remove the removal of the `master` taint, because kubeadm 1.25 will stop applying it.

Reply:
ic my confusion now, thanks